lib/usdUfe/ufe/UsdClipboardCommands.cpp

```diff
@@ -0,0 +1,496 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include "UsdClipboardCommands.h"
+
+#include <usdUfe/ufe/Global.h>
+#include <usdUfe/ufe/UsdUndoAddNewPrimCommand.h>
+#include <usdUfe/ufe/UsdUndoDuplicateSelectionCommand.h>
+#include <usdUfe/ufe/Utils.h>
+#include <usdUfe/undo/UsdUndoBlock.h>
+
+#include <pxr/usd/usd/stage.h>
+#include <pxr/usd/usdShade/material.h>
+#include <pxr/usd/usdShade/nodeGraph.h>
+#include <pxr/usd/usdShade/shader.h>
+
+#include <ufe/runTimeMgr.h>
+#include <ufe/sceneItemOps.h>
+
+namespace {
+
+constexpr std::string_view kClipboardMetadata = "ClipboardMetadata";
+constexpr std::string_view kMaterialName = "materialName";
+constexpr std::string_view kNodeName = "shaderName";
+constexpr std::string_view kStagePath = "stagePath";
+
+///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Utility functions
+void setClipboardMetadata(
+    UsdSceneItem::Ptr& item,
+    const std::string& metadataKey,
+    const std::string& metadataValue)
+{
+    const auto setMetadataCmd = item->setGroupMetadataCmd(
+        std::string(kClipboardMetadata), metadataKey, Ufe::Value(metadataValue));
+
+    if (setMetadataCmd) {
+        setMetadataCmd->execute();
+    }
+}
+
+void setClipboardMetadata(const UsdUndoDuplicateSelectionCommand::Ptr& duplicateSelectionCmd)
+{
+    for (const auto& duplicatedItem : duplicateSelectionCmd->getDuplicatedItemsMap()) {
+        const auto prim = duplicatedItem.first->prim();
+
+        // Do not set ClipboardMetadata for Materials.
+        if (PXR_NS::UsdShadeMaterial(prim)) {
+            continue;
+        }
+
+        std::string       materialName;
+        const std::string stagePath = duplicatedItem.first->path().popSegment().string();
+        const std::string nodeName = duplicatedItem.first->nodeName();
+
+        const auto parentMaterial = getParentMaterial(duplicatedItem.first);
+
+        // Recursively find the parent material.
+        if (parentMaterial) {
+            materialName = parentMaterial->nodeName();
+        }
+
+        // Get the usdItem for the copied item in the clipboard.
+        Ufe::PathSegment segment(
+            "/" + duplicatedItem.second->prim().GetName().GetString(),
+            UsdUfe::getUsdRunTimeId(),
+            '/');
+        auto usdItem = UsdSceneItem::create(segment, duplicatedItem.second->prim());
+
+        // Set the node name in the ClipboardMetadata.
+        setClipboardMetadata(usdItem, std::string(kNodeName), nodeName);
```

Comment:
Are these metadata and their purpose documented somewhere? They seem somewhat unrelated to copy/paste of USD data, but if they exist they must serve a purpose that should be documented.

Reply:
Original code from LookdevX - @alicedegirolamo can you help me with the doc for these. From my understanding of the code they are used to recreate the same structure (material with original name) when copy/pasting shaders.

Reply:
Thanks for the explanation Alice. I've added some extra comments for these metadata in the cpp file.